Add WebIDL sequence return values

emilio opened the pull request:
And use it for `WebGLRenderingContext::getSupportedExtensions`.
Part of #544

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/8412)

<!-- Reviewable:end -->

emilio committed:
Implement WebIDL sequence return values

emilio committed:
webgl: Implement getSupportedExtensions

emilio committed:
webgl: Add test for getSupportedExtensions

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

ecoal95 commented:
r? @jdm

ecoal95 commented on components/script/dom/bindings/conversions.rs:
```diff
@@ -806,6 +807,28 @@ impl<T: FromJSValConvertible> FromJSValConvertible for Option<T> {
+    fn to_jsval(&self, cx: *mut JSContext, rval: MutableHandleValue) {
+        unsafe {
+            let js_objects: Vec<JSVal> = self.iter().map(|obj| {
+                let mut val = mem::uninitialized();
```
Not 100% sure about if this is correct, I tested it and it worked though.

jdm commented:
-S-awaiting-review +S-needs-code-changes 
<!-- Reviewable comment -K2bq_7FQhpTesEyaRNf:1458266033 -->

---

Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
There are a couple unsafe things about, this, unfortunately. If `obj.to_jsval` triggers a GC, all of the previous elements that have been been will become dangling pointers since they're not rooted. We should do this instead:

```rust
let mut js_objects = RootedVec::new();
for obj in &self {
    let mut val = RootedValue::new(cx, UndefinedValue());
    obj.to_jsval(cx, val.handle_mut());
    js_objects.push(val.handle().get());
}
```
This will ensure that each converted value is stored in a rooted vector, so subsequent conversions will not risk GCing them.
<!-- Reviewable comment -K2bn4nhFwl9J-H8nywI:2105700195 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 827 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bpCFpgEQ_kKCVorjm)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L827)):</sup>
ObjectValue, since we've already checked the presence of null.
<!-- Reviewable comment -K2bpCFpgEQ_kKCVorjn:-1332814764 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 814 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bgvafxycDLLc0CgSs-r3-814)** ([raw file](https://github.com/servo/servo/blob/82c243a63f6d99e43885636b22c4f99ca89b24d7/components/script/dom/bindings/conversions.rs#L814)):</sup>
See below :) from_marked_location is a huge red flag, in general, since it means "I swear that the address I'm passing here is known to the GC".
<!-- Reviewable comment -K2boWXglrgTgat39MR4:576544824 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

ecoal95 commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
Trying this I get an error (`JSVal` is not traceable). I guess I can use a plain `Vec<RootedValue>` (the values are rooted so there should be no problem), and later get the handles?

```rust
            let js_objects: Vec<RootedValue> = self.iter().map(|obj| {                      
                let mut val = RootedValue::new(cx, UndefinedValue());
                obj.to_jsval(cx, val.handle_mut());
                val
            }).collect();                                                                   
            
            let values: Vec<JSVal> = js_objects.iter()                                      
                                               .map(|obj| obj.handle().get())               
                                               .collect(); 
```

I think this is correct (even though I'm not too happy with the double vector, but I guess we should keep the `RootedValues` while using the `JSValue`s), am I right?
<!-- Reviewable comment -K2bwsmdyGaWYiNceQWR:1308302060 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 814 \[r3\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bgvafxycDLLc0CgSs-r3-814)** ([raw file](https://github.com/servo/servo/blob/82c243a63f6d99e43885636b22c4f99ca89b24d7/components/script/dom/bindings/conversions.rs#L814)):</sup>
Thanks!
I supposed there would be issues with GC, I'm not familiar with the SpiderMonkey API.
<!-- Reviewable comment -K2c-2E9GYC7aETLch5x:-1638669117 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

ecoal95 commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
My fault, this crashes.
<!-- Reviewable comment -K2c2Hr4t8LhFwJ_H-Hb:1903221588 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 827 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bpCFpgEQ_kKCVorjm)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L827)):</sup>
Done.
<!-- Reviewable comment -K2c0vzEWGmfCQzQueoP:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

ecoal95 commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 817 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2bn4ngoGnu7aa9hW3S)** ([raw file](https://github.com/servo/servo/blob/3ef89a47c74ab18654aa1a6f7cc25d2c7a316564/components/script/dom/bindings/conversions.rs#L817)):</sup>
I ended up using `Heap<JSValue>`, and it works fine (passes the test adding a few strings to `getSupportedExtensions`).

May yo review it to see if there might be any issue with GC, or if this can be done in a more performant way? :)

TL;DR; here is the relevant code:

```rust
let mut js_objects = RootedVec::new();
 
for obj in self.iter() {
    let mut heap = Heap::default();
    let mut val = RootedValue::new(cx, UndefinedValue());
    obj.to_jsval(cx, val.handle_mut());
    heap.set(val.handle().get());
    js_objects.push(heap);
}
 
let values: Vec<JSVal> = js_objects.iter()
                                   .map(|heap| heap.handle().get())
                                   .collect();
<!-- Reviewable comment -K2c6vkzMM7WJETK27DJ:-844316405 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

Ms2ger commented:
I wonder if we should have a `fn push(&mut self, Handle<T>)` for `RootedVec<Heap<T>>`. @nox?

jdm commented:
-S-awaiting-review +S-needs-code-changes +S-awaiting-answer 
<!-- Reviewable comment -K2oT2jYmB9lB-4WAxQ2:1089396853 -->

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oQrAIxHapd9WYkl81)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L820)):</sup>
I think the order here should be:
```rust
let mut val = RootedValue::new(cx, UndefinedValue());
obj.to_jsval(cx, val.handle_mut());
js_objects.push(Heap::default());
js_objects.iter_mut().last().unwrap().set(val.handle().get());
```
<!-- Reviewable comment -K2oQrAIxHapd9WYkl82:-1383030984 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 825 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oSkdTcZB3xOTp9amv)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L825)):</sup>
The fact that we need to create a new vector with bare Value types gives me pause. I think we need to sort out whether we should be using `RootedVec<Value>` or `RootedVec<Heap<Value>>` before I can finish this review.
<!-- Reviewable comment -K2oSkdTcZB3xOTp9amw:-716999637 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 836 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oRmECIvJw7RVczbHO)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L836)):</sup>
`&*js_array` instead of transmuting, please.
<!-- Reviewable comment -K2oRmECIvJw7RVczbHP:-1450594437 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #8197) made this pull request unmergeable. Please resolve the merge conflicts.

emilio committed:
Implement WebIDL sequence return value using `JS_NewArrayObject1`

This commits changes the creation of the array to use the length, and
sets the elements later.

Reference implementation:
https://github.com/mozilla/gecko-dev/blob/07ed19e582faf40f5f72567169db62452f6c6dc9/js/xpconnect/public/nsTArrayHelpers.h#L16

ecoal95 commented:
Review status: 4 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oQrAIxHapd9WYkl81)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L820)):</sup>
I changed the implementation to take rid of this, using `JS_NewArrayObject1` (the overload with predefined length instead of values) and `JS_DefineElement` (to set numeric properties).

This also solves the double vector issue, and the `RootedVec`/`Heap` stuff.
<!-- Reviewable comment -K2rXN868RyfeSk6KqOG:1567623628 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 825 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oSkdTcZB3xOTp9amv)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L825)):</sup>
See above.
<!-- Reviewable comment -K2rXtPzIFCbw0rTSFWj:-1582418045 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 836 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oRmECIvJw7RVczbHO)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L836)):</sup>
Done.
<!-- Reviewable comment -K2rXvlC4TDUUlHrj6Zp:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

jdm commented:
Great; only nits remaining!
-S-needs-rebase -S-awaiting-review -S-awaiting-answer +S-needs-code-changes 
<!-- Reviewable comment -K2r_1JLqipQSNWEwNxk:859767431 -->

---

Reviewed 3 of 3 files at r5.
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r4\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2oQrAIxHapd9WYkl81)** ([raw file](https://github.com/servo/servo/blob/b19267acef977b8ff051464e7a21e397c8adb233/components/script/dom/bindings/conversions.rs#L820)):</sup>
Great solution :)
<!-- Reviewable comment -K2rZxDxyKMSKBFcWUNM:969955230 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 43 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rYxOh_Tf-JPCeO3u8)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L43)):</sup>
js::JSPROP_ENUMERATE
<!-- Reviewable comment -K2rYxOh_Tf-JPCeO3u9:2083333823 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZQzumQpK4c7hS4km)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L820)):</sup>
No need for `get()` here; `Handle<T>` dereferences to `&T`.
<!-- Reviewable comment -K2rZQzvVuIuZ82nLufM:-1449933027 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 822 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZ3iZdFu5LAggIq05)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L822)):</sup>
for (index, obj) in self.iter().enumerate() {
<!-- Reviewable comment -K2rZ3iZdFu5LAggIq06:1755120518 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

jdm commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

ecoal95 commented:
Review status: 6 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

<sup>**[components/script/dom/bindings/conversions.rs, line 820 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZQzumQpK4c7hS4km)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L820)):</sup>
Done.
<!-- Reviewable comment -K2ra9NjNkyxONGP8Vuy:-499146009 -->

---

<sup>**[components/script/dom/bindings/conversions.rs, line 822 \[r5\]](https://reviewable.io:443/reviews/servo/servo/8412#-K2rZ3iZdFu5LAggIq05)** ([raw file](https://github.com/servo/servo/blob/1f00c14e2810fa8b4871b056d104ed4944b901d8/components/script/dom/bindings/conversions.rs#L822)):</sup>
Good catch (sorry, I have little time and I tried to do it as quickly as possible).
<!-- Reviewable comment -K2ra0una6ss3xE062T7:-358016575 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

jdm commented:
@bors-servo: r+
<!-- Reviewable comment -K2rbVDDwMAxCaW53hyw:-317921516 -->

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8412)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit e33330d has been approved by `jdm`

<!-- @bors-servo r=jdm e33330db4eb9d9187c98139aad732ff86ab9d828 -->

bors-servo commented:
:hourglass: Testing commit e33330d with merge 12f6ba2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/417)

jdm commented:
@bors-servo: retry
* #8468

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/381), [gonk](http://build.servo.org/builders/gonk/builds/393), [linux-dev](http://build.servo.org/builders/linux-dev/builds/382), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/397), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/392), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/410) are reusable. Rebuilding only [linux-rel](http://build.servo.org/builders/linux-rel/builds/417)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/381), [gonk](http://build.servo.org/builders/gonk/builds/393), [linux-dev](http://build.servo.org/builders/linux-dev/builds/382), [linux-rel](http://build.servo.org/builders/linux-rel/builds/419), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/397), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/392), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/410)
